server/src/main/java/org/elasticsearch/TransportVersion.java

```diff
@@ -0,0 +1,292 @@
+
+    private static final TreeSet<TransportVersion> declaredVersions;
+
+    static {
+        Map<Integer, TransportVersion> builder = new HashMap<>();
+        Map<String, TransportVersion> uniqueIds = new HashMap<>();
+
+        Set<String> ignore = Set.of("ZERO", "CURRENT", "MINIMUM_COMPATIBLE");
+        for (Field declaredField : TransportVersion.class.getFields()) {
+            if (declaredField.getType().equals(TransportVersion.class)) {
+                String fieldName = declaredField.getName();
+                if (ignore.contains(fieldName)) {
+                    continue;
+                }
+                try {
+                    TransportVersion version = (TransportVersion) declaredField.get(null);
+
+                    TransportVersion maybePrevious = builder.put(version.id, version);
+                    assert maybePrevious == null
+                        : "expected [" + version.id + "] to be uniquely mapped but saw [" + maybePrevious + "] and [" + version + "]";
+
+                    TransportVersion sameUniqueId = uniqueIds.put(version.uniqueId, version);
+                    assert sameUniqueId == null
+                        : "Versions "
+                            + version
+                            + " and "
+                            + sameUniqueId
+                            + " have the same unique id. Each TransportVersion should have a different unique id";
+
+                } catch (IllegalAccessException e) {
+                    assert false : "Version field [" + fieldName + "] should be public";
+                }
+            }
+        }
+        idToVersion = Map.copyOf(builder);
+        declaredVersions = idToVersion.values().stream().collect(Collectors.toCollection(() -> new TreeSet<>()));
+    }
+
+    public static TransportVersion readVersion(StreamInput in) throws IOException {
+        return fromId(in.readVInt());
+    }
+
+    public static TransportVersion fromId(int id) {
+        TransportVersion known = idToVersion.get(id);
+        if (known != null) {
+            return known;
+        }
+        // this is a version we don't otherwise know about - just create a placeholder
+        return new TransportVersion(id, "<unknown>");
+    }
+
+    public static void writeVersion(TransportVersion version, StreamOutput out) throws IOException {
+        out.writeVInt(version.id);
+    }
+
+    /**
+     * Returns the minimum version between the 2.
+     */
+    public static TransportVersion min(TransportVersion version1, TransportVersion version2) {
+        return version1.id < version2.id ? version1 : version2;
+    }
+
+    /**
+     * Returns the maximum version between the 2
+     */
+    public static TransportVersion max(TransportVersion version1, TransportVersion version2) {
+        return version1.id > version2.id ? version1 : version2;
+    }
+
+    // TODO for testing or a getter? should this be used?
```

Comment:
The purpose of this TODO is not clear, can it be expanded on or removed?

Reply:
Removed